pennylane/ops/op_math/prod.py

```diff
@@ -0,0 +1,267 @@
+
+    Args:
+        factors (tuple[~.operation.Operator]): a tuple of operators which will be multiplied together.
+
+    Keyword Args:
+        do_queue (bool): determines if the product operator will be queued (currently not supported).
+            Default is True.
+        id (str or None): id for the product operator. Default is None.
+
+    **Example**
+
+    >>> prop_op = Prod(qml.PauliX(0), qml.PauliZ(0))
+    >>> prop_op
+    PauliX(wires=[0]) @ PauliZ(wires=[0])
+    >>> qml.matrix(prop_op)
+    array([[ 0,  -1],
+           [ 1,   0]])
+    >>> prop_op.terms()
+    ([1.0], [PauliX(wires=[0]) @ PauliZ(wires=[0])])
+
+    .. details::
+        :title: Usage Details
+
+        The Prod operator represents both matrix composition and tensor products
+        between operators.
+
+        >>> prod_op = Prod(qml.RZ(1.23, wires=0), qml.PauliX(wires=0), qml.PauliZ(wires=1))
+        >>> prod_op.matrix()
+        array([[ 0.        +0.j        ,  0.        +0.j        ,
+                 0.81677345-0.57695852j,  0.        +0.j        ],
+               [ 0.        +0.j        ,  0.        +0.j        ,
+                 0.        +0.j        , -0.81677345+0.57695852j],
+               [ 0.81677345+0.57695852j,  0.        +0.j        ,
+                 0.        +0.j        ,  0.        +0.j        ],
+               [ 0.        +0.j        , -0.81677345-0.57695852j,
+                 0.        +0.j        ,  0.        +0.j        ]])
+    """
+    _name = "Prod"
+    _eigs = {}  # cache eigen vectors and values like in qml.Hermitian
+
+    def __init__(self, *factors, do_queue=True, id=None):  # pylint: disable=super-init-not-called
+        """Initialize a Prod instance"""
+        self._id = id
+        self.queue_idx = None
+
+        if len(factors) < 2:
+            raise ValueError(f"Require at least two operators to multiply; got {len(factors)}")
+
+        self.factors = factors
+        self._wires = qml.wires.Wires.all_wires([f.wires for f in self.factors])
+
+        if do_queue:
+            self.queue()
+
+    def __repr__(self):
+        """Constructor-call-like representation."""
+        return " @ ".join([f"{f}" for f in self.factors])
+
+    def __copy__(self):
+        cls = self.__class__
+        copied_op = cls.__new__(cls)
+        copied_op.factors = tuple(copy(f) for f in self.factors)
+
+        for attr, value in vars(self).items():
+            if attr not in {"factors"}:
+                setattr(copied_op, attr, value)
+
+        return copied_op
+
+    def terms(self):  # is this method necessary for this class?
+        return [1.0], [self]
+
+    @property
+    def data(self):
+        """Create data property"""
+        return [f.parameters for f in self.factors]
+
+    @data.setter
+    def data(self, new_data):
+        """Set the data property"""
+        for new_entry, op in zip(new_data, self.factors):
+            op.data = copy(new_entry)
```

Comment:
Why a copy?

Reply:
<img width="1125" alt="image" src="https://user-images.githubusercontent.com/21964864/182438036-611106ac-9627-46b3-97ec-70a3607a95ba.png">

Reply:
yes I think this case is the example? Arithmetic operations need to keep track of the data parameters for each of their summands / factors. In this case we store the data for each factor (which is a list in itself) as a separate entry in the data property instead of concatenating them so that we can easily recover which parameters correspond to which observable without keeping track of indices.

Reply:
What you show was exactly the behaviour I expected, and it is consistent with how our operations work. For example,

```
op = qml.RX(1.0, wires=0)
new_data = [2.0]
op.data = new_data
new_data[0] = 3.0
print(op.data)
```

Reply:
Maybe in the future, we can consider making `data` a tuple instead of a list if we want to forbid this type of manipulations.

Reply:
Wait, so we want this aliasing? This is not very good practice...

Reply:
This is behaviour that we would love to change, and remains partly due to historic decisions. I think making the tape immutable will also allow us to finally make the operations immutable

Reply:
For now, copying is an extra step we don't necessarily need to be doing.

The behaviour without copying is consistent with how we treat `Operator` data.  `op.data` is usually accessed via `op.parameters` anyway, which is a copy of the data and can't be set. If we add a `copy` here, we should add a copy *everywhere* data is set.

I personally would support making `Operator`'s immutable and getting rid of the distinction between `data` and `parameters`, but that is a question for another time.

Reply:
Ahh okay sounds good, as long as we are aware this aliasing exists